### PR TITLE
8264634: CollectCLDClosure collects duplicated CLDs when dumping dynamic archive

### PR DESCRIPTION
**Pull request by kelthuzadx**
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/qingfeng.yy/openjdk16_so_warning/jdk/src/hotspot/share/classfile/classLoaderData.cpp:316), pid=68929, tid=68930
#  assert(_keep_alive > 0) failed: Invalid keep alive decrement count
#
# JRE version: OpenJDK Runtime Environment (17.0) (slowdebug build 17-internal+0-adhoc.qingfengyy.jdk)
# Java VM: OpenJDK 64-Bit Server VM (slowdebug 17-internal+0-adhoc.qingfengyy.jdk, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x781087]  ClassLoaderData::dec_keep_alive()+0x31

Stack: [0x00007f1593072000,0x00007f1593173000],  sp=0x00007f1593171c00,  free space=1023k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x781087]  ClassLoaderData::dec_keep_alive()+0x31
V  [libjvm.so+0xef19e7]  MetaspaceShared::link_and_cleanup_shared_classes(Thread*)+0x181
V  [libjvm.so+0x1260834]  JavaThread::invoke_shutdown_hooks()+0x46
V  [libjvm.so+0x12609e5]  Threads::destroy_vm()+0xe7
V  [libjvm.so+0xbb40ec]  jni_DestroyJavaVM_inner+0x91
V  [libjvm.so+0xbb4147]  jni_DestroyJavaVM+0x1f
C  [libjli.so+0x4b4f]  JavaMain+0xc61
C  [libjli.so+0xad93]  ThreadJavaMain+0x27
```
We observed VM crashed when dumping dynamic archive in a simple springboot application(See detailed content on JBS attachment). I did some investigations. In rare case, both of the following paths may be stepped on when dumping dynamic archive:

```
1. SIGINT
at java.lang.Shutdown.beforeHalt(java.base@17-internal/Native Method)
at java.lang.Shutdown.exit(java.base@17-internal/Shutdown.java:172)
- locked <0x00000007fef02040> (a java.lang.Class for java.lang.Shutdown)
at java.lang.Terminator$1.handle(java.base@17-internal/Terminator.java:51)
at jdk.internal.misc.Signal$1.run(java.base@17-internal/Signal.java:219)
at java.lang.Thread.run(java.base@17-internal/Thread.java:831)

2. Normal Exit
JavaThread::invoke_shutdown_hooks()+0x46
Threads::destroy_vm()+0xe7
jni_DestroyJavaVM_inner+0x91
jni_DestroyJavaVM+0x1f
JavaMain+0xc61
ThreadJavaMain+0x27
```

They would call MetaspaceShared::link_and_cleanup_shared_classes, and CollectCLDClosure collects duplicated CLDs into _loaded_cld, _keep_alive is decrementing twice, causing a negative _keep_alive.

Testing(linux_x64):
[+] test/hotspot/jtreg/runtime/cds
[+] test/hotspot/jtreg/gc

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264634](https://bugs.openjdk.java.net/browse/JDK-8264634): CollectCLDClosure collects duplicated CLDs when dumping dynamic archive

### Reviewers
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3320/head:pull/3320` \
`$ git checkout pull/3320`

Update a local copy of the PR: \
`$ git checkout pull/3320` \
`$ git pull https://git.openjdk.java.net/jdk pull/3320/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3320`

View PR using the GUI difftool: \
`$ git pr show -t 3320`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3320.diff">https://git.openjdk.java.net/jdk/pull/3320.diff</a>

</details>

### CODE REVIEW

**Review by yminqi — COMMENTED** [review]
Hi, Yi
  The _loaded_cld is a global list, in this case it looks contain duplicated CLD in it. 
   The duplication could from the thread run shutdown hook.
   Could you try 
  if (!cld->is_unloading()) {
      cld->inc_keep_alive();
+     if (!_loaded_cld->contains(cld)) {
      _loaded_cld->append(cld);
+  }
    }
Please let us know if you can avoid the crash.

File: src/hotspot/share/memory/metaspaceShared.cpp [hunk 1]
```diff
@@ -592,10 +593,11 @@ void MetaspaceShared::link_and_cleanup_shared_classes(TRAPS) {
     // Keep scanning until we have linked no more classes.
   }
 
-  for (int i = 0; i < _loaded_cld->length(); i++) {
-    ClassLoaderData* cld = _loaded_cld->at(i);
+  for (int i = 0; i < loaded_cld.length(); i++) {
+    ClassLoaderData* cld = loaded_cld.at(i);
     cld->dec_keep_alive();
   }
+  loaded_cld.trunc_to(0);
```
Comment (iklam) [inline src/hotspot/share/memory/metaspaceShared.cpp:600]: There's no need for the trucate -- `loaded_cld` is locally allocated and will be freed after this function returns.

Also, to improve modularity, I think we should move the dec_keep_alive loop into the destructor of CollectCLDClosure.

Also, `loaded_cld` can be moved as a field into CollectCLDClosure.

File: src/hotspot/share/memory/metaspaceShared.cpp [hunk 2]
```diff
@@ -565,16 +564,18 @@ bool MetaspaceShared::link_class_for_cds(InstanceKlass* ik, TRAPS) {
 
 void MetaspaceShared::link_and_cleanup_shared_classes(TRAPS) {
   // Collect all loaded ClassLoaderData.
-  CollectCLDClosure collect_cld;
+  ResourceMark rm;
+  GrowableArray<ClassLoaderData*> loaded_cld;
+  CollectCLDClosure collect_cld(&loaded_cld);
```
Comment (iklam) [inline src/hotspot/share/memory/metaspaceShared.cpp:569]: I think we should add a comment to say why it's necessary to first collect the ClassLoaderDatas first:

```
// ClassLoaderDataGraph::loaded_cld_do requires ClassLoaderDataGraph_lock.
// We cannot link the classes while holding this lock (or else we may run into deadlock).
// Therefore, we need to first collect all the CLDs, and then link their classes after
// releasing the lock.
```

**Review by iklam — CHANGES_REQUESTED** [review]
The fix looks reasonable. If MetaspaceShared::link_and_cleanup_shared_classes may be called twice, it's better to isolate the loaded_cld for each invocation. Allocating it locally will also avoid any potential threading issues.

I have some requests for cleaning up the code.

**Comment by kelthuzadx** [comment]
> Hi, Yi
> The _loaded_cld is a global list, in this case it looks contain duplicated CLD in it.
> The duplication could from the thread run shutdown hook.
> Could you try
> if (!cld->is_unloading()) {
> cld->inc_keep_alive();
> `+` if (!_loaded_cld->contains(cld)) {
> _loaded_cld->append(cld);
> `+` }
> }
> Please let us know if you can avoid the crash.

Hi Yumin, this fix still crashes because the CLDs collected at the first invocation of MetaspaceShared::link_and_cleanup_shared_classes are not cleaned, they will decrement their _keep_alives as before at the second invocation of MetaspaceShared::link_and_cleanup_shared_classes.

**Comment by kelthuzadx** [comment]
Hi Ioi,

> Also, to improve modularity, I think we should move the dec_keep_alive loop into the destructor of CollectCLDClosure.
> Also, loaded_cld can be moved as a field into CollectCLDClosure.

Suggestions make sense, changed. Tests under runtime/cds/ are all passed with slowdebug mode.

**Review by iklam — APPROVED** [review]

**Review by yminqi — APPROVED** [review]
Make the CLD list local is a reasonable solution. LGTM.

**Comment by kelthuzadx** [comment]
Thanks @yminqi @iklam for the reviews!

/integreate

**Pull request closed without merging**